### PR TITLE
Challenge/sort numbers in ascending order

### LINKED ISSUE #12: Ascending Order Arranger (Number)
| **Field**        | **Details**                                                                                                         |
|------------------|---------------------------------------------------------------------------------------------------------------------|
| **Name**         | New Challenge                                                                                                       |
| **Description**        | Write a program that sorts a list of numbers from smallest to largest.                                 |
| **Title**        |        Numbers Sorter in ascending order                                                                                                 |

### PR DESCRIPTION
**Pull request by Safiya-hash**
Wrote a function to sort numbers in ascending order using bubble sort strategy

Resolves https://github.com/MIT-Emerging-Talent/ET6-foundations-group-13/issues/12

Here's a review checklist:---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: collaboration/group_norms.md [hunk 1]
```diff
@@ -17,8 +17,13 @@
 
 3. **Supportive Environment**
 
+   - We will foster an atmosphere where personal challenges that affect  
+   - attendance can
+   be openly shared.
+
    - We will foster an atmosphere where personal challenges that affect
    attendance can be openly shared.
+
```
Comment (tosan-okome) [inline collaboration/group_norms.md]: @Safiya-hash Here's the source of the failed CI check. There are trailing spaces on this line and Markdown doesn't like that. Simply deleting the extra spaces should fix it.

File: solutions/sort.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+Module contents:
+    - sort: A function to sort a list of integers in ascending order.
+
+Created on 03 January 2025
+@author: Safiya Hash
+
+"""
+
+
+def sort(numbers: list[int]) -> list[int]:
```
Comment (tosan-okome) [inline solutions/sort.py]: Just a suggestion. How about naming the function "sort_ascending"  or "sort_asc" since it only sorts in ascending order?

File: solutions/sort.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    """
+    The function takes a list of int & sorts it in ascending order using the bubble Sort strategy.
+        numbers (list[int]): This is the list of numbers we want to sort
+        list[int]: A sorted list of integers from smallest to largest
+        numbers == [-float("inf"), 0, float("inf")]
```
Comment (tosan-okome) [inline solutions/sort.py]: Lines 16 - 18 should be under "Parameters".

File: solutions/tests/test_sort.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    def test_large_range_int_list(self):
+        """it should return the list of large integers from smallest to largest"""
+        self.assertEqual(
+            sort([35236, 555555, 98732145, 100, 7543, 894521, 99991200]),
+            [100, 7543, 35236, 555555, 894521, 98732145, 99991200],
+        )
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (tosan-okome) [inline solutions/tests/test_sort.py]: I'm not sure I understand what these two lines of code (92 -93) are. Can you enlighten me?

**Review by tosan-okome — COMMENTED** [review]
@Safiya-hash Your code is very well documented and you have a lot of unit tests which is a good thing. Well done! 

I highlighted a few minor suggestions (and a question) in the files. Do let me know what you think and we can proceed with a merge. 😃

**Comment by tosan-okome** [comment]
@Safiya-hash I also noticed that you used force push twice. Did you encounter conflicts somewhere?

**Comment by Safiya-hash** [comment]
> @Safiya-hash I also noticed that you used force push twice. Did you encounter conflicts somewhere?

Yes, I encountered conflict and mistakenly merged main branch of ET6/foundation with my branch. Then I used force push second time to revert my commit.

File: collaboration/group_norms.md (hunk 1 above)
Comment (Safiya-hash) [reply collaboration/group_norms.md]: Thanks, I am updating it.

File: solutions/sort.py (hunk 2 above)
Comment (Safiya-hash) [reply solutions/sort.py]: Yeah, it is a good suggestion. I will change it to sort_ascending.

File: solutions/sort.py (hunk 3 above)
Comment (Safiya-hash) [reply solutions/sort.py]: Good catch, I will fix it.

File: solutions/tests/test_sort.py (hunk 4 above)
Comment (Safiya-hash) [reply solutions/tests/test_sort.py]: The lines ensure that the unit tests in the script are executed only when the script is run directly, not when it is imported as a module, however, maybe it is not needed if we follow the doc and testing examples, so I am removing it.

Comment (tosan-okome) [reply solutions/tests/test_sort.py]: Oh. I see. 
I'm not exactly sure what that means 😄 but I can do some further research on my own. Thanks.

Comment (Safiya-hash) [reply solutions/tests/test_sort.py]: I was just using it to run the unittests without using the terminal but as it wasn't a necessity I removed it. Terminal is always there to check and run our tests so no worries.

**Comment by Safiya-hash** [comment]
> @Safiya-hash Your code is very well documented and you have a lot of unit tests which is a good thing. Well done!
> 
> I highlighted a few minor suggestions (and a question) in the files. Do let me know what you think and we can proceed with a merge. 😃

**Comment by Safiya-hash** [comment]
> @Safiya-hash Your code is very well documented and you have a lot of unit tests which is a good thing. Well done!
> 
> I highlighted a few minor suggestions (and a question) in the files. Do let me know what you think and we can proceed with a merge. 😃

Thank you for your kind review ❤️

**Review by tosan-okome — APPROVED** [review]
@Safiya-hash Everything looks good now. I'm proceeding with a merge. Congratulations! 🎉
